File: June 2022/June_21.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        int i = 0, n = heights.size();
        priority_queue<int, vector<int>, greater<int> > pq;
        while(++i < n && pq.size() < ladders){
            int diff = heights[i] - heights[i - 1];
            if(diff > 0)
                pq.push(diff);
        }
        i--;
        while(++i < n){
            int diff = heights[i] - heights[i - 1];
            if(diff <= 0)
                continue;
            if(!pq.empty() && pq.top() < diff && bricks >= pq.top()){
                bricks -= pq.top();
                pq.pop();
                pq.push(diff);
            } else if(bricks >= diff)
                bricks -= diff;
            else
                return i - 1;
        }
        return n - 1;
    }
};

File: June 2022/June_21.java
class Solution {
    public int furthestBuilding(int[] heights, int bricks, int ladders) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int i = 1;
        for(; pq.size() < ladders && i < heights.length; i++)
            if(heights[i] > heights[i - 1])
                pq.add(heights[i] - heights[i - 1]);
        for(; i < heights.length; i++){
            int diff = heights[i] - heights[i - 1];
            if(diff <= 0)
                continue;
            if(!pq.isEmpty() && pq.peek() < diff && bricks >= pq.peek()){
                bricks -= pq.remove();
                pq.add(diff);
            } else if(bricks >= diff)
                bricks -= diff;
            else
                return i - 1;
        }
        return heights.length - 1;
    }
}
